File: 2020/10.py
adapters = [0]
with open("10.txt") as file:
    for line in file:
        adapters.append(int(line.strip()))

adapters.sort()
adapters.append(adapters[-1] + 3)
differences = {"1": 0, "3": 0}
for i in range(len(adapters) - 1):
    if abs(adapters[i] - adapters[i+1]) == 1:
        differences["1"] += 1
    elif abs(adapters[i] - adapters[i+1]) == 3:
        differences["3"] += 1
print(differences["1"]*differences["3"])

paths = {}
for i in range(-1, adapters[-1] + 1):
    paths[i] = 0
paths[0] = 1
paths[1] = 1

for i in range(2, len(adapters)):
    paths[adapters[i]] = paths[adapters[i] - 1] + paths[adapters[i] - 2] + paths[adapters[i] - 3]
print(paths[adapters[-1]])

File: 2020/Lua/10.lua
local nums, pres = {0}, {[0] = true}
for line in io.lines("10.txt") do
  table.insert(nums, tonumber(line))
  pres[tonumber(line)] = true
end

table.sort(nums)
table.insert(nums, nums[#nums] + 3)
pres[nums[#nums]] = true

local dif1, dif3 = 0, 0
for _, n in pairs(nums) do
  if pres[n+1] then
    dif1 = dif1 + 1
  elseif pres[n+3] then
    dif3 = dif3 + 1
  end
end
print(dif1 * dif3)

local ways = {[0] = 1}
for _, n in pairs(nums) do
  if n ~= 0 then
    ways[n] = (ways[n-3] or 0) + (ways[n-2] or 0) + (ways[n-1] or 0)
  end
end
print(ways[nums[#nums]])
